test/kubernetes_unittest.cc

```diff
@@ -1215,4 +1215,153 @@ TEST_F(KubernetesTestFakeServer, MetadataQuery) {
   EXPECT_EQ(pod_metadata->ToString(), m[3].metadata().metadata->ToString());
 }
 
+class KubernetesTestFakeServerOneWatchRetry : public KubernetesTest {
+ protected:
+  void SetUp() override {
+    server.reset(new testing::FakeServer());
+    KubernetesTest::SetUp();
+  }
+
+  std::unique_ptr<Configuration> CreateConfig() override {
+    return std::unique_ptr<Configuration>(
+      new Configuration(std::istringstream(
+        "InstanceId: TestID\n"
+        "InstanceResourceType: gce_instance\n"
+        "InstanceZone: TestZone\n"
+        "KubernetesClusterLocation: TestClusterLocation\n"
+        "KubernetesClusterName: TestClusterName\n"
+        "KubernetesEndpointHost: " + server->GetUrl() + "\n"
+        "KubernetesNodeName: TestNodeName\n"
+        "MetadataIngestionRawContentVersion: TestVersion\n"
+        "KubernetesUpdaterWatchConnectionRetries: 1\n"
+        "KubernetesUseWatch: true\n"
+      )));
+  }
+
+  std::unique_ptr<testing::FakeServer> server;
+};
+
+// Polls store until collected_at for resource is newer than
+// last_timestamp.  Returns false if newer timestamp not found after 3
+// seconds (polling every 100 millis).
+bool WaitForNewerCollectionTimestamp(const MetadataStore& store,
+                                     const MonitoredResource& resource,
+                                     Timestamp last_timestamp) {
+  for (int i = 0; i < 30; i++){
+    const auto metadata_map = store.GetMetadataMap();
+    const auto m = metadata_map.find(resource);
+    if (m != metadata_map.end() && m->second.collected_at > last_timestamp) {
+      return true;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  return false;
+}
+
+TEST_F(KubernetesTestFakeServerOneWatchRetry, KubernetesUpdater) {
+  // Set responses for fake server representing the Kubernetes master.
+  server->SetResponse("/api/v1/nodes?limit=1", "{}");
+  server->SetResponse("/api/v1/pods?limit=1", "{}");
+  server->AllowStream(
```

Comment:
[Optional] What you have now is fine, but I wonder if a better model would be to create a watch handler object that you associate with a path, so the API would look something like:
```c++
  auto* pod_watch_handler = server->CreateWatchHandler(
      "/api/v1/pods?fieldSelector=spec.nodeName%3DTestNodeName&watch=true");
  auto* node_watch_handler = server->CreateWatchHandler(
      "/api/v1/watch/nodes/TestNodeName?watch=true");
  ...
  EXPECT_TRUE(pod_watch_handler->WaitForWatcher(time::seconds(3)));
  EXPECT_TRUE(node_watch_handler->WaitForWatcher(time::seconds(3)));
  ...
  node_watch_handler->SendStreamResponse(node_metadata->ToString());
  ...
  pod_watch_handler->SendStreamResponse(pod_metadata->ToString());
```
It's not a huge refactoring, but I would not want to predicate approving this PR on it. Thoughts?

Reply:
Yeah, that would work too. Let's leave as is for this PR and I'll consider it for a future PR. Thanks.

Reply:
[Optional] Does it make sense to add a TODO?